File: product-catalog-service/src/main/java/com/carlosoliveira/ecommerce/productcatalog/domain/Money.java
package com.carlosoliveira.ecommerce.productcatalog.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import lombok.*;

import java.math.BigDecimal;

@Embeddable
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@EqualsAndHashCode
public class Money {
    
    @Column(nullable = false, precision = 19, scale = 4)
    private BigDecimal amount;

    @Column(nullable = false, length = 3)
    private Currency currency;

    public Money add(Money other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException("Cannot add Money with different currencies.");
        }
        return new Money(this.amount.add(other.amount), this.currency);
    }

    public Money subtract(Money other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException("Cannot subtract Money with different currencies.");
        }
        return new Money(this.amount.subtract(other.amount), this.currency);
    }

    public Money multiply(BigDecimal multiplier) {
        return new Money(this.amount.multiply(multiplier), this.currency);
    }
}

File: shopping-cart-service/src/domain/value-objects/money.ts
import { ValueObject } from './value-object';

export class Money extends ValueObject<{ amount: number; currency: string }> {
  static zero(): Money {
    return new Money({ amount: 0, currency: 'USD' });
  }

  add(other: Money): Money {
    this.ensureSameCurrency(other);
    return new Money({
      amount: this.value.amount + other.value.amount,
      currency: this.value.currency,
    });
  }

  multiply(factor: number): Money {
    return new Money({
      amount: this.value.amount * factor,
      currency: this.value.currency,
    });
  }

  private ensureSameCurrency(other: Money): void {
    if (this.value.currency !== other.value.currency) {
      throw new Error(
        `Currency mismatch: cannot operate on ${this.value.currency} and ${other.value.currency}`,
      );
    }
  }
}
